daemon/daemon_unix.go

```diff
@@ -131,20 +132,23 @@ func (daemon *Daemon) adaptContainerSettings(hostConfig *runconfig.HostConfig) {
 // verifyPlatformContainerSettings performs platform-specific validation of the
 // hostconfig and config structures.
 func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *runconfig.HostConfig, config *runconfig.Config) ([]string, error) {
-	var warnings []string
+	warnings := []string{}
+	sysInfo := sysinfo.New(false)
 
 	if hostConfig.LxcConf.Len() > 0 && !strings.Contains(daemon.ExecutionDriver().Name(), "lxc") {
 		return warnings, fmt.Errorf("Cannot use --lxc-conf with execdriver: %s", daemon.ExecutionDriver().Name())
 	}
+
+	// memory subsystem checks and adjustments
 	if hostConfig.Memory != 0 && hostConfig.Memory < 4194304 {
 		return warnings, fmt.Errorf("Minimum memory limit allowed is 4MB")
 	}
-	if hostConfig.Memory > 0 && !daemon.SystemConfig().MemoryLimit {
+	if hostConfig.Memory > 0 && !sysInfo.MemoryLimit {
 		warnings = append(warnings, "Your kernel does not support memory limit capabilities. Limitation discarded.")
 		logrus.Warnf("Your kernel does not support memory limit capabilities. Limitation discarded.")
 		hostConfig.Memory = 0
```

Comment:
That's one solution for the problem I mentioned.
```
$ docker run --name ttt -m 300M busybox true
$ sudo umount /sys/fs/cgroup/memory
$ docker start ttt
```
In this scenario, after docker run, we'll have hostConfig.Memory = 300M and hostConfig.MemorySwap = 600M, then we umount memory cgroup and try to start container, `sysinfo.MemoryLimit` would be false, so we set `hostConfig.Memory = 0`, but then we'll get `hostConfig.Memory == 0 && hostConfig.MemorySwap > 0` and end up error.

So here when we set `hostConfig.Memory = 0` we can also set `hostConfig.MemorySwap = -1`, it'll fix it, and because if we don't limit memory, we can't limit memory swap either, so the logic is right.